api/src/main/java/org/openmrs/module/pihcore/htmlformentry/action/AddPatientToQueueAction.java

```diff
@@ -0,0 +1,104 @@
+ */
+public class AddPatientToQueueAction implements CustomFormSubmissionAction {
+
+    private static final String ADD_TO_QUEUE_CONCEPT_CODE_PIH = "1272";
+    protected Log log = LogFactory.getLog(getClass());
+
+    @Override
+    public void applyAction(FormEntrySession formEntrySession) {
+
+        Concept addToQueueConcept = Context.getConceptService().getConceptByMapping(ADD_TO_QUEUE_CONCEPT_CODE_PIH, "PIH");
+
+        if (addToQueueConcept == null) {
+            throw new RuntimeException("AddPatientToQueueAction not executing as Concept PIH: " + ADD_TO_QUEUE_CONCEPT_CODE_PIH + " cannot be found");
+        }
+        Patient patient = formEntrySession.getPatient();
+        Encounter encounter = formEntrySession.getEncounter();
+
+        for (Obs candidate : encounter.getObsAtTopLevel(false)) {
+            if (candidate.getConcept().equals(addToQueueConcept)) {
+                // The coded answer represents the Concept associated with the queue service
+                Concept serviceQueue = candidate.getValueCoded();
+                QueueSearchCriteria criteria = new QueueSearchCriteria();
+                criteria.setServices(Arrays.asList(serviceQueue));
+                QueueServicesWrapper queueServicesWrapper = Context.getRegisteredComponents(QueueServicesWrapper.class).get(0);
+                List<Queue> queues = queueServicesWrapper.getQueueService().getQueues(criteria);
+                Queue queue = null ;
+                QueueEntry queueEntry = null;
+                if ( queues.isEmpty() ) {
+                    throw new RuntimeException("AddPatientToQueueAction not executing as no queue with service concept " + serviceQueue.getName() + " could be found ");
+                }
+                if ( queues.size() > 1 ) {
+                    throw new RuntimeException("AddPatientToQueueAction not executing as there are more than one queue with service concept " + serviceQueue.getName());
+                }
+                queue = queues.get(0);
+
+                queueEntry = new QueueEntry();
+                queueEntry.setQueue(queue);
+                queueEntry.setPatient(patient);
+
+                List<Concept> allowedStatuses = queueServicesWrapper.getAllowedStatuses(queue);
+                if (!allowedStatuses.isEmpty()) {
+                    queueEntry.setStatus(allowedStatuses.get(0));
+                }
+
+                List<Concept> allowedPriorities = queueServicesWrapper.getAllowedPriorities(queue);
+                if (!allowedPriorities.isEmpty()) {
+                    queueEntry.setPriority(allowedPriorities.get(0));
+                }
+                queueEntry.setStartedAt(encounter.getEncounterDatetime());
+
+                QueueEntrySearchCriteria searchCriteria = new QueueEntrySearchCriteria();
+                searchCriteria.setPatient(patient);
+                searchCriteria.setQueues(Collections.singletonList(queue));
```

Comment:
Tangentially, it would probably have been useful to have added an additional property to QueueEntrySearchCriteria (Date activeOnDate), to only return those queue entries from the DB, rather than have to iterate on the like we are doing below, but this is fine.